### PR TITLE
Painless: Plumb the definition through CompilerSettings

### PR DESCRIPTION
**Pull request by nik9000**
Plumbs the Definition through the CompilerSettings rather than on its own. This seems a bit more natural to me, though I'm not 100% sure about it.

Also fixes some weirdness leftover from the Definition migration. Still doesn't add support for more than a single definition.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/CompilerSettings.java [hunk 1]
```diff
@@ -137,4 +142,18 @@ public boolean areRegexesEnabled() {
...
+     * The whitelist to compile against.
+     */
+    public Definition getDefinition() {
+        return definition;
+    }
+
+    /**
+     * Set the whitelist to compile against.
+     */
+    public void setDefinition(Definition definition) {
```
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/CompilerSettings.java]: This *feels* like the more natural place to configure the whitelist, though I'm not sure about it.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java [hunk 2]
```diff
@@ -518,15 +517,15 @@ public RuntimeClass getRuntimeClass(Class<?> clazz) {
     private final Map<String, Struct> structsMap;
     private final Map<String, Type> simpleTypesMap;
 
-    private Definition() {
+    private Definition(List<String> definitionFiles) {
```
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: I'm not sure this is how this'll end up, but it is closer to what we want in the end I think.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java [hunk 3]
```diff
@@ -490,26 +489,26 @@ public Struct getStruct() {
 
     /** Returns whether or not a non-array type exists. */
     public boolean isSimpleType(final String name) {
-        return BUILTINS.structsMap.containsKey(name);
```
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: These were just wrong.

**Comment by rjernst** [comment]
I don't think this is the right path. First, I don't see this as a compiler setting. It is something that is required of compiling, I think it should just be an argument to compile. Second, having Definition ctor take the list of files won't work long term. We need this to be an intermediate structure which contains eg a Map of Class and then methods/handles/whatever for each class that should be whitelisted. Otherwise Definition here would try loading the classes using Class.forName, which would assume the class is accessible from the painless classloader, which it might not be (for example, with another plugin registering a script context along with classes available). I also think the builtins (minus the actual types that _must_ exist like primitive types, HashMap, etc) will need to be moved out for painless to be useable on its own.  For it to be useful as an embeddable language, it should assume nothing is accessible except that required by the syntax for the language.

**Pull request closed without merging**